executor/analyze.go

```diff
@@ -116,148 +103,40 @@ func (e *AnalyzeExec) Next(ctx context.Context, req *chunk.Chunk) error {
-
-	pruneMode := variable.PartitionPruneMode(e.ctx.GetSessionVars().PartitionPruneMode.Load())
-	// needGlobalStats used to indicate whether we should merge the partition-level stats to global-level stats.
-	needGlobalStats := pruneMode == variable.Dynamic
-	type globalStatsKey struct {
-		tableID int64
-		indexID int64
-	}
-	type globalStatsInfo struct {
-		isIndex int
-		// When the `isIndex == 0`, histIDs will be the column IDs.
-		// Otherwise, histIDs will only contain the index ID.
-		histIDs      []int64
-		statsVersion int
-	}
-	// globalStatsMap is a map used to store which partition tables and the corresponding indexes need global-level stats.
-	// The meaning of key in map is the structure that used to store the tableID and indexID.
-	// The meaning of value in map is some additional information needed to build global-level stats.
-	globalStatsMap := make(map[globalStatsKey]globalStatsInfo)
-	for panicCnt < concurrency {
-		results, ok := <-resultsCh
-		if !ok {
-			break
-		}
-		if results.Err != nil {
-			err = results.Err
-			if isAnalyzeWorkerPanic(err) {
-				panicCnt++
-			} else {
-				logutil.Logger(ctx).Error("analyze failed", zap.Error(err))
-			}
-			finishJobWithLog(e.ctx, results.Job, err)
-			continue
-		}
-		if results.TableID.IsPartitionTable() && needGlobalStats {
-			for _, result := range results.Ars {
-				if result.IsIndex == 0 {
-					// If it does not belong to the statistics of index, we need to set it to -1 to distinguish.
-					globalStatsID := globalStatsKey{tableID: results.TableID.TableID, indexID: int64(-1)}
-					histIDs := make([]int64, 0, len(result.Hist))
-					for _, hg := range result.Hist {
-						// It's normal virtual column, skip.
-						if hg == nil {
-							continue
-						}
-						histIDs = append(histIDs, hg.ID)
-					}
-					globalStatsMap[globalStatsID] = globalStatsInfo{isIndex: result.IsIndex, histIDs: histIDs, statsVersion: results.StatsVer}
-				} else {
-					for _, hg := range result.Hist {
-						globalStatsID := globalStatsKey{tableID: results.TableID.TableID, indexID: hg.ID}
-						globalStatsMap[globalStatsID] = globalStatsInfo{isIndex: result.IsIndex, histIDs: []int64{hg.ID}, statsVersion: results.StatsVer}
-					}
-				}
-			}
-		}
-		if err1 := statsHandle.SaveTableStatsToStorage(results, results.TableID.IsPartitionTable()); err1 != nil {
-			err = err1
-			logutil.Logger(ctx).Error("save table stats to storage failed", zap.Error(err))
-			finishJobWithLog(e.ctx, results.Job, err)
-		} else {
-			finishJobWithLog(e.ctx, results.Job, nil)
-			// Dump stats to historical storage.
-			if err := e.recordHistoricalStats(results.TableID.TableID); err != nil {
-				logutil.BgLogger().Error("record historical stats failed", zap.Error(err))
-			}
-		}
-	}
 	for _, task := range e.tasks {
 		if task.colExec != nil && task.colExec.memTracker != nil {
 			task.colExec.memTracker.Detach()
 		}
 	}
+	pruneMode := variable.PartitionPruneMode(e.ctx.GetSessionVars().PartitionPruneMode.Load())
```

Comment:
Maybe we need to switch the order between `memTracker.Detach` and `handleResultsError`. We wait for results coming from `resultsCh` in `handleResultsError` and before that, some of  `e.tasks` may be still running.

Reply:
updated